### ISSUE TITLE
[Ready for payment via newdot][$16000] Android - Copy / Paste / Cut menu is not displayed when selecting text in compose box - Reported by: @parasharrajat

### ISSUE THREAD

**Issue by kidroca** [Expensify/App#8349]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Navigate to a conversation
2. Type anything in compose box
3. Select it

## Expected Result:
Native menu to cut / copy / paste should be displayed

## Actual Result:
Native menu to cut / copy / paste is not displayed or displayed for a second and then hidden

## Workaround:
User is unable to cut / copy / paste text in the compose box.
Selecting a word/text and then long tapping on another word seems to bring the context menu with copy/cut/paste actions

## Additional details

This [comment from the previous ticket](https://github.com/Expensify/App/issues/6876#issuecomment-1078047007) summarizes what we have found out and tried so far.

We've identified using the `selection` prop works as expected in iOS, but does not work as expected on Android, resulting in the reported bug.

We're looking for identifying the problem in `react-native` and proposing a fix
Once a proposal is approved a contributor would be hired to submit a PR against our react-native fork: https://github.com/Expensify/react-native

We're **not** looking for a workaround solution that
- uses platform specific code e.g. handling `selection`  differently in `Composer/index.android|ios|web.js` files
- uses `setTimeout`, `setImmediate` or `requestAnimationFrame`

## Platform:

Where is this issue occurring?
- Android

**Version Number:** v1.1.46-3
**Reproducible in staging?:** yes
**Reproducible in production?:** yes 
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://user-images.githubusercontent.com/44479856/147117514-155c964a-587d-4581-82d6-5b842041902a.mp4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @parasharrajat
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1640179428463500
**Previous ticket:** https://github.com/Expensify/App/issues/6876

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by kidroca** [comment]
cc @mallenexpensify @parasharrajat @rushatgabhane @stitesExpensify

**Comment by greg-schroeder** [comment]
I reproduced this one so sending to my friends in Eng

**Comment by Julesssss** [comment]
I assigned @parasharrajat and made the issue the same price as the original, as this is not a simple fix.

**Comment by NicMendonca** [comment]
Upwork job post: https://www.upwork.com/jobs/~0191b930fd2b38bae8

**Comment by mallenexpensify** [comment]
Doubled price to $2000
https://www.upwork.com/jobs/~0191b930fd2b38bae8

**Comment by mallenexpensify** [comment]
Took over as Contributor Manager (CM) too as I was assigned to the previous issue

**Comment by mallenexpensify** [comment]
Doubled price to $4000
https://www.upwork.com/jobs/~0191b930fd2b38bae8

**Comment by mallenexpensify** [comment]
Doubled price to $8000
https://www.upwork.com/jobs/~0191b930fd2b38bae8

**Comment by mallenexpensify** [comment]
Any reason I shouldn't double this @Julesssss @parasharrajat ?

**Comment by mallenexpensify** [comment]
Good idea @Julesssss , I created [this internal SO](https://stackoverflow.com/c/expensify/questions/12797) asking were good posts are to post these jobs
> Sometimes we have jobs with high-dollar prices and it'd be helpful to promote them outside of #expensify-open-source, E/App and Upwork. Where are some places to promote these jobs? (ie. reddit, discord, GH repos, sites, ???)

**Comment by metehanozyurt** [comment]
I think i found problem can i still make proposal @mallenexpensify ? . Thanks🙏

**Comment by metehanozyurt** [comment]
I search commits and research. 
cause of this problem about this issue
https://github.com/Expensify/App/issues/2601

If we add this line our problem solved.
https://github.com/Expensify/App/blob/32d8f0261b825cb6e47e441b9cf08400340a164e/src/components/Composer/index.android.js#L66-L76
because of `selection` props cant open the first time.
if we change like this other platforms not effected and android works fine.
```js
<RNTextInput
                autoComplete="off"
                placeholderTextColor={themeColors.placeholderText}
                ref={el => this.textInput = el}
                maxHeight={CONST.COMPOSER_MAX_HEIGHT}
                rejectResponderTermination={false}
                editable={!this.props.isDisabled}
                /* eslint-disable-next-line react/jsx-props-no-spreading */
                {...this.props}
                selection={undefined}
            />
```

I am taking video right now .

**Comment by mdneyazahmad** [comment]
## Proposal

Removing this line fixes the issue.

https://github.com/Expensify/App/blob/fe432c30fdd46db9c34d714c64db78c496761b7c/src/pages/home/report/ReportActionCompose.js#L550

```diff
                                    isDisabled={isComposeDisabled || isBlockedFromConcierge}
-                                   selection={this.state.selection}
                                    onSelectionChange={this.onSelectionChange}
```

We are omitting this prop from `ios` therefore it works there and this issue exists only on `android`.  We need to also remove these two lines.

https://github.com/Expensify/App/blob/fe432c30fdd46db9c34d714c64db78c496761b7c/src/components/Composer/index.ios.js#L75-L76

### Result:
![Screenshot_1651227815](https://user-images.githubusercontent.com/77761491/165927590-8016aa2d-105b-4ae4-ac84-ab100ccb6dd4.png)

**Comment by mdneyazahmad** [comment]
@parasharrajat I know there are other issues related to `selection` prop, but in our case I think we need not explicitly pass `selection` prop. We do need selection values to insert emojis and we already have `onSelectionChange` that will update when selection changes.

As we can see that it works fine without this prop on `ios`  so I think we can also do the same on other platform. In case I am missing something. I would appreciate it, if you tell me what I am missing here.

Thank you!

**Comment by Julesssss** [comment]
Not overdue, still awaiting proposals

**Comment by mallenexpensify** [comment]
@Julesssss @parasharrajat any reason we _shouldn't_ double this?

**Comment by Julesssss** [comment]
> @Julesssss @parasharrajat any reason we _shouldn't_ double this?

Go for it

**Comment by mallenexpensify** [comment]
Doubled to $16k! https://www.upwork.com/jobs/~015051a36f053c1362

**Comment by ahmdshrif** [comment]
# **Proposal** 

## Root couse 
I debug this by creating a native android app and trying to `setSelction` on EdiatText. 
and as I expect setSelction on native does not show the context menu by default. 

the copy/paste menu only appears when the selection is triggered by os itself. but if we trigger setSelction manually it will only select the text as the behavior on the video above. and the reason for that. some apps have their custom copy past menus like this one. 

![P3yAN](https://user-images.githubusercontent.com/21364901/170833862-8e5c307f-e38c-44c7-93c1-ddf4d42eb136.jpg)


but the question was why that happens on our case and setSelction on this case not manage by our code it comes from Os itself  . so after debugging the logic on react native code they use these steps to track the selection change from native.

 
![Screen Shot 2022-05-28 at 6 12 05 PM](https://user-images.githubusercontent.com/21364901/170834257-d3214d00-abcb-42f9-9708-8a09aab224a2.png)


the conclusion here is the native send to the app there is change on selection . and then app set it on the state . and when the state change the native will set the selection again and set selection will interrupt the context menu appearance 


## Sloution 

we only need an intermediate selection state so we will not pass the selection props that come from the app directly instead we will pass this intermediate section one    
and we will compare it to the native state and if it is equal we will not update it. because it will rewrite the selection and the context menu will disappears . 



i try this with rn-teset app and all  tests is work fine 

and context menu appears when selecting the  text as expected (when using selection state ) 

changed we need 
https://github.com/ahmdshrif/react-native/commit/fb02f08fe1733aa27b92443386e72034d0a55c61



## Result 
https://user-images.githubusercontent.com/21364901/170835044-3783ad86-8bfe-46ab-98a5-d63eb3416100.mov



by the way, we can use setSelction  on the app to workaround this issue. 


cc @parasharrajat

**Comment by ahmdshrif** [comment]
i just do more test and the above proposal have a big issue so for not waste your time ... please ignore it.

**Comment by ahmdshrif** [comment]
@parasharrajat @Julesssss  i update the [proposal](https://github.com/Expensify/App/issues/8349#issuecomment-1140484993) now 
sorry for making it confusing.

**Comment by kidroca** [comment]
> 3- set the selection to null if it's not changed to not rewrite the native state.
> 
> ```ts
>          selection={isSelectionChange ? selection : null}
> ```

If I remember correctly this would not work on ios and require a workaround to skip the fix for that platform 

@Julesssss 
It's not exactly a bug in App - we're using the `selection` prop to set initial selection / cursor position, and because we're passing this prop we have to update it whenever the selection changes, otherwise it would remain  with the initial value
(Imagine using the value prop, you have to keep track of it in state and pass it as it changes otherwise the field would be stuck with whatever you've passed for value. This is called a controlled component and it causes a rerender for each character change)

**Comment by kidroca** [comment]
> I note react-native set selection to null if they do not need it to update the native state.

Where do they do that
- in their examples on how to use text fields /  selection 
- or internally for some special handling?

Couldn't we apply this fix to set selection to null if the value is already the same in react native android?

I guess some additional logic is triggered when set native props is called, because echoing the same selection back should just do nothing 

Maybe setting to null causes the field to recompute selection internally - can we find a way to properly reset internal values without forcing this recomputation via a null value

**Comment by ahmdshrif** [comment]
Thanks, @kidroca for your feedback 
but it looks like my proposal was not clear to you. 

I see from comments you think I do these changes on the app level (we can do this ) 
but I do these changes on the bridge level instead. 
and the bridge has a different component for android than ios. 

please check here https://github.com/ahmdshrif/react-native/commit/9f8bf5003a5f62b44b03a980654f83c2fe105f4f

why not on native android? 

if we register a state it will automatically update the component's native state when it receives changes from the native 
it's not like the imperative way (ref.setSelction() ) wich. call function on native we can change on it. 
it's some generic code from react-native that can change the native state based on change on the js side. 
and any change on it will affect all components and all state.  (that's what I find from some debug  but not found any doc explaining what happen exactly if you have something that can help will be fine .) 

react-native provides the 2 options (use props or use imperative way (setSeleciton) ) 

we can also use `setSelciton`  in case we receive selection props and stop sending selection as props to native. 
and on this case, we can control when we should do an update or not. 
the changes in this case will be: - 

on bridge : - 
```ts
useEffect(()=>{
if (selection != nativeSelciton || nativeSelction == null){
ref.setSelciotn(selciton)
}
},[selection,nativeSelciton])

```

or add the condition to native as you suggest but we still need the useeffect without the condition. 

```java 

seSelciton (start, end) {
if (start == this.lastStart && end == this.lastEnd) {
return 
}
//.... other code to calc real start and real end 

super.setSelciton(start,end) 
}
``` 


but I don't know if react native will accept delete props or not.  so I go with the first proposal I mention (keep it but make undefined when no changes) 


we can also use setSelciton on the app level and delete the selection props and this will solve the issue for us. 

Anyway, the proposal is to stop the duplication. I prefer to do this on the bridge 
and that needs to stop updating props when it's the same as a native one. we can do this by setSelciton or by adding conditions on props as I explain. 

the same way can add to the app also if we need to make it simple. 
or to native with imperative if we need to go deeper. 

and the setSelciton will work on both android and ios without this issue.

**Comment by marktoman** [comment]
The problem is caused by [this line](https://github.com/facebook/react-native/blob/ba6bf5a3ce7039a7e407a6632ee41aa3d504f833/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L567) in RN Android.

### Proposal

Wrap the line with the below to prevent redundant calls:
```java
if (!(getText().toString()).equals(reactTextUpdate.getText().toString())) {
  getText().replace(0, length(), spannableStringBuilder);
}
```

We can add a similar guard to [setSelection](https://github.com/facebook/react-native/blob/ba6bf5a3ce7039a7e407a6632ee41aa3d504f833/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L354) given it fires redundantly as well, but this does not have any effect on the bug.
```java
if (super.getSelectionStart() != start || super.getSelectionEnd() != end) {
  super.setSelection(start, end);
}
```

**Comment by ahmdshrif** [comment]
> Wrap the line with the below to prevent redundant calls:
> ```java
> if (!(getText().toString()).equals(reactTextUpdate.getText().toString())) {
>   getText().replace(0, length(), spannableStringBuilder);
> }
> ```

You broken the text change and not solve the issue at all. 
I recommend to test all test cases in rn app with any chance you make . 



> We can add a similar guard to [setSelection](https://github.com/facebook/react-native/blob/ba6bf5a3ce7039a7e407a6632ee41aa3d504f833/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactEditText.java#L354) given it fires redundantly as well, but this does not have any effect on the bug.
> ```java
> if (super.getSelectionStart() != start || super.getSelectionEnd() != end) {
>   super.setSelection(start, end);
> }
> ```

This true and i mention we can do this on my last comment.

**Comment by kidroca** [comment]
@ahmdshrif 

There's already one workaround code that deal with selection here: https://github.com/ahmdshrif/react-native/blob/9f8bf5003a5f62b44b03a980654f83c2fe105f4f/Libraries/Components/TextInput/TextInput.js#L932-L942

```js
  // Android sends a "onTextChanged" event followed by a "onSelectionChanged" event, for
  // the same "most recent event count".
  // For controlled selection, that means that immediately after text is updated,
  // a controlled component will pass in the *previous* selection, even if the controlled
  // component didn't mean to modify the selection at all.
  // Therefore, we ignore selections and pass them through until the selection event has
  // been sent.
  // Note that this mitigation is NOT needed for Fabric.
  // discovered when upgrading react-hooks
  // eslint-disable-next-line react-hooks/exhaustive-deps
  let selection: ?Selection =
```

Seems your proposal and the existing code around `lastNativeSelectionState` are very similar,
both compare `selection` to the last captured native selection: https://github.com/ahmdshrif/react-native/blob/9f8bf5003a5f62b44b03a980654f83c2fe105f4f/Libraries/Components/TextInput/TextInput.js#L996-L1004

```js
    if (
      selection &&
      lastNativeSelection &&
      (lastNativeSelection.start !== selection.start ||
        lastNativeSelection.end !== selection.end)
    ) {
      nativeUpdate.selection = selection;
      setLastNativeSelection({selection, mostRecentEventCount});
    }
```

Passing `null` seems like a simpler solution and I wonder why they didn't go with it initially but instead went with using `mostRecentEventCount` approach

From the logic in the code comment above it seems the idea is to pass `selection` even if it's the same as the `nativeSelection`, and they are only passing `null` until the event count settles, e.g. they avoid passing `selection` for a brief time, the are not trying to pass `null` when the native selection is already the same as the `selection` prop

**Comment by ahmdshrif** [comment]
```ts
 if (
      selection &&
      lastNativeSelection &&
      (lastNativeSelection.start !== selection.start ||
        lastNativeSelection.end !== selection.end)
    ) {
      nativeUpdate.selection = selection;
      setLastNativeSelection({selection, mostRecentEventCount});
    }
```

i take some idea form this  existing code  but at the end the `nativeUpdate.selection` and `setLastNativeSelection` 
not affect the props that send to native it is just some value they save to calculate where they should send an event to native with ` viewCommands.setTextAndSelection()`. 

but the selection props is sent every time that it gets updated so i want to do something similar to props to not update it if not change.
and I explain some ways we can do this above.

>From the logic in the code comment above it seems the idea is to pass selection even if it's the same as the nativeSelection, and they are only passing null until the event count settles, e.g. they avoid passing selection for a brief time, the are not trying to pass null when the native selection is already the same as the selection prop

if you do not send `null`  and native get update to a different value I try this at first but when you use `setText` the curser will get stuck between two position.  we need to fix this and any case similar if we decided to go wit this direction . 

I write this solution at being and when I get this issue mention it here. https://github.com/Expensify/App/issues/8349#issuecomment-1140494039 and update the proposal.

**Comment by ahmdshrif** [comment]
@kidroca 

it calls this setter function 
https://github.com/ahmdshrif/react-native/blob/9f8bf5003a5f62b44b03a980654f83c2fe105f4f/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputShadowNode.java#L196-L206

```ts 
  @ReactProp(name = PROP_SELECTION)
  public void setSelection(@Nullable ReadableMap selection) {
    mSelectionStart = mSelectionEnd = UNSET;
    if (selection == null) return;


    if (selection.hasKey("start") && selection.hasKey("end")) {
      mSelectionStart = selection.getInt("start");
      mSelectionEnd = selection.getInt("end");
      markUpdated();
    }
  }
```

if we send a selection with null nothing will happen the code will return early . and react-native already uses it in the bridge as you mention. 

why not send a return also if the selection is the same as the previous one . ?

I try this but the shadow node will update when it receives any props with `  @ReactProp ` 
the only way to make it not update is to send null because react-native tells this is what happens when the props are not defined on js. 

>in case when setter is of some complex type, null will be provided as a default value in case when corresponding property gets removed.

https://reactnative.dev/docs/native-components-android#3-expose-view-property-setters-using-reactprop-or-reactpropgroup-annotation


the props that receive on viewManger can be easy to handle than this .
just return if it the same to previos one  bt i note react-native team move it from view manager to shadow node to fix a crash happen in specific case. 

from that i think the only way now is to set null as they do.


@parasharrajat  what do you think .?

**Comment by parasharrajat** [comment]
Hmm, I have to agree that I am a little behind on this. I will read the whole discussion and follow up.

**Comment by kidroca** [comment]
@ahmdshrif 

> it calls this setter function https://github.com/ahmdshrif/react-native/blob/9f8bf5003a5f62b44b03a980654f83c2fe105f4f/ReactAndroid/src/main/java/com/facebook/react/views/textinput/ReactTextInputShadowNode.java#L196-L206
> 
> ```ts
>   @ReactProp(name = PROP_SELECTION)
>   public void setSelection(@Nullable ReadableMap selection) {
>     mSelectionStart = mSelectionEnd = UNSET;
>     if (selection == null) return;
> 
> 
>     if (selection.hasKey("start") && selection.hasKey("end")) {
>       mSelectionStart = selection.getInt("start");
>       mSelectionEnd = selection.getInt("end");
>       markUpdated();
>     }
>   }
> ```

So it seems like the call to `markUpdated()` causes some layout/render side effect that we try to avoid by passing `null` instead of the unchanged selection

I think it'll be better to be handled here where it'll be obvious that we're trying to skip that behavior when `selection` didn't change instead of passing `null` and explaining with comments that we have to do it so the code here can exit early 

```java
 @ReactProp(name = PROP_SELECTION)
 public void setSelection(@Nullable ReadableMap selection) {
   if (selection == null) return;
   if (!selection.hasKey("start") || !selection.hasKey("end")) return;

   int start = selection.getInt("start");
   int end = selection.getInt("end");
   
   // we don't want to trigger `markUpdated` when selection didn't really change
   if (start == mSelectionStart && end == mSelectionEnd) return;

   mSelectionStart = start;
   mSelectionEnd = end;
   markUpdated();
}
```

And perhaps this would allow us to get rid of the core code that sometimes passes `null` for the same reason

**Comment by ahmdshrif** [comment]
@kidroca  
logically that should work. 
but let's explain what happens when I debug this. 

I do the flowing test: -

1-  keep function empty  **=> not work**
```java
@ReactProp(name = PROP_SELECTION)
 public void setSelection(@Nullable ReadableMap selection) {

}
```

2- put the selection to unset  **=> not work** 
```java
@ReactProp(name = PROP_SELECTION)
 public void setSelection(@Nullable ReadableMap selection) {
  mSelectionStart = mSelectionEnd = UNSET;
  return;
}
```


3 - delete the function it self  **=> work fine **
but change selciton  form state will not effect native 
```java
//@ReactProp(name = PROP_SELECTION)
// public void setSelection(@Nullable ReadableMap selection) {
 // mSelectionStart = mSelectionEnd = UNSET;
 // return;
//}
```

4- move function to view manager and return if the selection is identical to previous one  **=> work fine **  (but we can't do this according to this pr https://github.com/facebook/react-native/pull/22723/files ) 

```java
@ReactProp(name = "selection")
  public void setSelection(ReactEditText view, @Nullable ReadableMap selection) {
    if (selection == null |)) {
      return;
    }

    if (selection.hasKey("start") && selection.hasKey("end")) {
      if (selection.hasKey("start")  == view.getSelcitnStart() && selection.hasKey("end")  == view.getSelcitnEnd()){
      return ; 
      } 
      view.setSelection(selection.getInt("start"), selection.getInt("end"));
    }
  }
```


from this test, I go to this result. https://github.com/Expensify/App/issues/8349#issuecomment-1148303905

and I think this is why the react-native team uses null also.

**Comment by kidroca** [comment]
@ahmdshrif 

So what you're saying is that even if the `setSelection` method is empty and always returns, the bug with the context menu still happens (it gets interrupted and doesn't appear)?

**Comment by ahmdshrif** [comment]
@kidroca  yes

**Comment by ahmdshrif** [comment]
I also think about deleting the seter function from shadow node  (native) since in js when selection props state changes we send it also to native by event emitter also.
`viewCommands.setTextAndSelection()`

and that looks like duplication.  to send by props and by an event at the same time. 
the event has duplication protected by some condition inside useEffect so it does not interrupt the context menu. but props will interrupt.

for me this work fine 

but again I think the safe way is to send null like the current code to not break something. Maybe this duplicate is a workaround for another issue.

**Comment by kidroca** [comment]
I guess the `@ReactProp` annotation triggers something internally and causes a re-render when the value passed to the annotated method is not the default value (default `selection` is `null`), so even if the method is empty something is still happening under the hood

**Comment by ahmdshrif** [comment]
@kidroca 
That's what i see also. 
And ftom that  sending null is something safe to do . And solve the isse  as the current code do . 

do you still think it's wrong.. and if yes why ?

**Comment by ahmdshrif** [comment]
it's great news that fabric will solve this issue and similar issues. 

but on the other hand, the new architecture is still beta. and I expect to be unstable like a flutter (in the first year) for while. 
and maybe some lib that we use not be compatible with it. so it can be very a long road. 

and since we have something fix the current code we should fix it until the new Architecture gets stable.  
and the third party we use gets compatible. and the app itself is ready for it.

**Comment by mallenexpensify** [comment]
@parasharrajat @Julesssss what to you think of @kidroca 's _proposal_?
https://github.com/Expensify/App/issues/8349#issuecomment-1151172897

**Comment by Julesssss** [comment]
Hi all, thanks again for all the proposals and discussion. @ahmdshrif while I agree that there may be some instability, it looks to me that Fabric is close to release. If released, it would certainly be the optimal choice. 

> Since Fabric is the future and the issue is resolved with it, we should not try to work on a fix for the legacy renderer, but make App meet the requirements to enable Fabric (update react, react-native and other libraries)

I agree with this, and think we should go with @kidroca's proposal. Though not quite the same scenario, we upgraded to Hermes early and saw nothing but improvements to the app. Let's live at the bleeding edge. 

The only suggestion I have before moving forward is that we share the Fabric proposal in the public Slack channel, just in case anyone calls out an exception. I'm sure everyone will be on board, but let's just run it by the rest of the team. Does that work for you @kidroca?

**Comment by parasharrajat** [comment]
We def want to upgrade to fabric. There was a discussion in past on this. Only challenge was that many of the third party libs does not support fabric yet. So transition will not be smooth.

**Comment by kidroca** [comment]
@Julesssss 

I thought of Hermes as well 😄 

> The only suggestion I have before moving forward is that we share the Fabric proposal in the public Slack channel, just in case anyone calls out an exception. I'm sure everyone will be on board, but let's just run it by the rest of the team. Does that work for you @kidroca?

I was actually planning to post about the "New RN Architecture" anyway because it literally replaces the bridge with Fabric and Turbo Modules, the bridge is the main RN bottle neck and source of some performance issues 
I'll definitely post on slack now

---

While switching on the new architecture is as easy as flipping a switch to `true` we'll indeed need to keep this switch to `false` until our dependencies are compatible 
I see this as a continuous work where we try to do everything necessary to make App work/build with Fabric and the new architecture
The good thing is that we can indeed achieve this gradually 
- update `react-native` to at least v0.68
- make a list of dependencies who already have newer versions that support fabric and we can update
    - use this list to Upgrade App
- make a list of dependencies who don't yet have a fabric update
    - use this list to help get those package to a Fabric compatible version
- flip the `newArchEnabled` switch to `true` when above steps are satisfied
- if it turns out it's still too unstable we can just flip the switch back, but at least we'll be ready to put it to `true` shortly

**Comment by Julesssss** [comment]
> because it literally replaces the bridge with Fabric and Turbo Modules, the bridge is the main RN bottle neck and source of some performance issues

Oh wow, that's great! Can't wait for this.

So we should probably create a new issue for the migration, and put this one on hold. We'll then consider this one fixed once we've migrated to Fabric successfully. Does that work for you?

**Comment by kidroca** [comment]
@Julesssss 

> So we should probably create a new issue for the migration, and put this one on hold. We'll then consider this one fixed once we've migrated to Fabric successfully. Does that work for you?

Sounds like a plan

**Comment by parasharrajat** [comment]
We already have a issue here https://github.com/Expensify/App/issues/8503. we can break this into two. 
1. Upgrade the app to latest RN version. It involves version upgrade and make sure that app still works. Upgrades other packages if necessary. 
2. Activate Fabric architecture, try to fix issues via package upgrades. List out the packages that are not yet compatible. More of an audit. If Fabric just works. That is a great news.

**Comment by kidroca** [comment]
@parasharrajat 

This would work, but I'm not sure we need a 2nd issue, the 2nd issue sounds like the one with the heavy lifting, while updating react to 0.68 should be trivial 
In any case the work would be staged and in multiple PRs

**Comment by parasharrajat** [comment]
> I guess related issues would be opened up once we audit and start helping on external projects

Correct. Once we have an audit report and we know the status, a series of new issues can be opened to cover them.

**Comment by ahmdshrif** [comment]
@Julesssss 

> while I agree that there may be some instability, it looks to me that Fabric is close to release. If released, it would certainly be the optimal choice.

I totally agree on upgrading and counting this as a fix for this issue, but I think preparing the app and making other 3 party support fabric takes a long time. 
some third parties that I work with them in the app do not have any updates for many years. so we should fork and support them who knows if the new arch will work fine with them or not. 

and as @parasharrajat  says : 
> We still want the app be fully usable at any stage.

I suggest we can go in 2 directions: -  

* short term:  fix the app for the current arch. which will work until we get ready (maybe months to a year )  

* long-term which upgrading to a new arch when everything gets ok (new arch get stable, third parties get ready, app get ready ).

**Comment by mallenexpensify** [comment]
Bumped to `Weekly` since, I think, we're awaiting proposals and we don't need daily updates (let me know if that's not the case)

**Comment by mallenexpensify** [comment]
Updated to `Monthly` since I don't _think_ anyone thinks the bug currently warrants a higher price tag.  Correct me if you think I'm wrong (please... any/everyone, not just employees)

**Comment by stitesExpensify** [comment]
I agree that this should not be increased any more.

**Comment by dave-amit** [comment]
> Upwork job post: https://www.upwork.com/jobs/~0191b930fd2b38bae8

link not working!

**Comment by dave-amit** [comment]
sorry, above msg sent by me by mistake, I didn't see collapsed lines!

**Comment by kadiealexander** [comment]
Not overdue, still seeking proposals!

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by kidroca** [comment]
@kadiealexander 
> Not overdue, still seeking proposals!

I don't think we're seeking new proposals, because we've found out that upgrading `react-native` to a version that uses the Fabric renderer would fix this issue: https://github.com/Expensify/App/issues/8349#issuecomment-1151172897

And it seems the upgrade landed in staging recently, hence the issue is not reproducible ☝️

**Comment by kidroca** [comment]
We might need to do some refactoring the `ReportActionCompose` because AFAIK `setNativeProps` shouldn't be used with Fabric

**Comment by kidroca** [comment]
Actually, the new architecture is not yet enabled in Android, and I was able to reproduce the bug: 

https://github.com/Expensify/App/blob/a0c91b91e4ec5e85adb9391bdc56eb36c53dec1e/android/gradle.properties#L39-L43

Anyway, I think we should pause proposals until we try the new architecture

**Comment by kadiealexander** [comment]
@kidroca forgive my ignorance, do you know roughly when the new architecture is coming?

**Comment by hannojg** [comment]
In this PR: https://github.com/Expensify/App/pull/11684 i completely got rid of the `selection` prop, and on my android devices the menu is being displayed. Can anyone of you who had the issue test the PR and can confirm its fixed?

**Comment by parasharrajat** [comment]
I will check.

**Comment by Julesssss** [comment]
Also testing this today

**Comment by Julesssss** [comment]
> In this PR: #11684 i completely got rid of the `selection` prop (functionality is of course preserved), and on my android devices the menu is being displayed. Can anyone of you who had the issue test the PR and can confirm it's fixed?

Confirmed that this PR resolved the issue on a Physical Nexus 5X device.

**Comment by parasharrajat** [comment]
Great. I will test it today and review the PR as well if that is needed.

**Comment by parasharrajat** [comment]
Update: PR https://github.com/Expensify/App/pull/11684 is reviwed and close to finish line.

**Comment by kadiealexander** [comment]
Removing the `help wanted` label as we're not looking for proposals (per the above comment).

**Comment by Julesssss** [comment]
Still held

**Comment by parasharrajat** [comment]
Basically, we don't have to hold this. https://github.com/Expensify/App/pull/11684 is fixing this issue directly. I am testing this issue on that PR as well as a few other cases including performance. 

![image](https://user-images.githubusercontent.com/24370807/202050600-9accb2a3-5e5b-401d-93bf-6f2a931138b3.png)

**Comment by parasharrajat** [comment]
Update: Reviewed the PR today but the tests are still incomplete. Hopefully, it will be ready to be merged tomorrow.

**Comment by kadiealexander** [comment]
@parasharrajat how are we tracking with this?

**Comment by parasharrajat** [comment]
There is a new issue on the PR that is being fixed.

**Comment by kadiealexander** [comment]
PR above is being actively worked on and reviewed.

**Comment by kadiealexander** [comment]
Same as above.

**Comment by kadiealexander** [comment]
@parasharrajat any updates on this one?

**Comment by hannojg** [comment]
It's in another round of testing review. Some light bugs were reported, which I will sort out today.

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @Julesssss, @parasharrajat, @kadiealexander eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by hannojg** [comment]
The PR against the Expensify react native repo is up:

- https://github.com/Expensify/react-native/pull/50

**Comment by s77rt** [comment]
Hi team!
We have this issue https://github.com/Expensify/App/issues/12854 where controlled selection is causing us some troubles in cursor position. I'm [suggesting](https://github.com/Expensify/App/issues/12854#issuecomment-1506213114) to make the Composer not selection controlled, this should also fix this issue. But now I see we have a RN PR and maybe that will fix our issue too.

@hannojg Can you confirm if that PR will fix https://github.com/Expensify/App/issues/12854 as well

**Comment by hannojg** [comment]
@s77rt It very likely will, I can allocate some time this week to update the PR and then test for the bug behaviour. If it fixes, I will let you know 😊

**Comment by s77rt** [comment]
@hannojg That would be great! Thanks

**Comment by s77rt** [comment]
Should be fixed by https://github.com/Expensify/App/pull/17687

https://user-images.githubusercontent.com/16493223/233202048-3e86713c-c732-4e1d-b389-7a2bb5a133ec.mp4

___

We are having a discussion here https://expensify.slack.com/archives/C01GTK53T8Q/p1681938841671029 about the use of setSelection.

**Comment by hannojg** [comment]
Just a quick update: team Margelo identified a native fix for that upstream in react native, we will post a proposal this week 😊

**Comment by janicduplessis** [comment]
Hello, we have identified the root cause of the issue and have a fix for it upstream in RN as well as our fork. See https://github.com/Expensify/react-native/pull/55 fro a complete explanation of the problem and the fix.

**Comment by janicduplessis** [comment]
https://github.com/facebook/react-native/commit/d4f6cf1d80b6ee725b81d9fdccc263d193178249 Just landed, I will check to get it cherry picked in 0.72. I also have an open PR against our fork if we don’t want to wait for 0.72.

**Comment by AndrewGable** [comment]
Going to assign @Julesssss to [review the Fork PR for 2 here](https://github.com/Expensify/App/issues/8349#issuecomment-1574359870)

**Comment by Julesssss** [comment]
Merged the PR, now I'm trying to figure out how to make the release: [S.O](https://stackoverflowteams.com/c/expensify/q/14023/297) + [Readme](https://github.com/Expensify/react-native#expensify-release)

**Comment by Julesssss** [comment]
Phew, this is not exactly simple. After reading through the docs I have a few questions, so asked for help [here](https://github.com/Expensify/react-native/pull/55#issuecomment-1578419062).

**Comment by kadiealexander** [comment]
Not overdue, [this](https://github.com/Expensify/Expensify/issues/289343) is still being worked on.

**Comment by janicduplessis** [comment]
The fix was cherry-picked into 0.72, and released as 0.72.0-rc.6 https://github.com/facebook/react-native/releases/tag/v0.72.0-rc.6

**Comment by janicduplessis** [comment]
@parasharrajat There was some discussion around stop using a fork recently since all changes have been merged upstream and maintaining it is complex. Not sure if a decision has been made thought, maybe @Julesssss @roryabraham knows?

**Comment by roryabraham** [comment]
Yes, we are abandoning the fork and focusing on getting stuff merged upstream.

**Comment by Julesssss** [comment]
So just to be clear, can someone confirm that we should hold this issue until either: 
- We merge the upstream changes into our fork
- We kill the fork

**Comment by Julesssss** [comment]
I'm afraid this was just discussed internally at the moment. I'll keep an eye out for the final outcome and will share the issue that hopefully will be created to track it.

**Comment by Julesssss** [comment]
Aha thanks. I completely forgot that I had reacted to it already 😅

**Comment by kadiealexander** [comment]
Hey team! I'm just adding a note that I'll be OOO until June 26th. Since this is on hold I won't reassign it, but if you need a bug0 teammate this week, please reapply the label! Thanks.

**Comment by kadiealexander** [comment]
@Julesssss @parasharrajat the PR that this was on hold for got closed, would one of you mind taking a look and seeing if we can be unblocked, or if there's more that needs to be done before we can get started on this?

**Comment by hannojg** [comment]
This issue is ultimately blocked by us trying to get rid of the react native fork, and trying to use the upstream version directly.
I don't think we currently have a tracking issue for this task @Julesssss ?

**Comment by hannojg** [comment]
Pardon, we do have an issue for that here

https://github.com/Expensify/App/issues/18444

Seems like for RN 0.72 we will still use the fork. We can hold this issue for the one named above

**Comment by Julesssss** [comment]
Updated the hold, thanks!

**Comment by janicduplessis** [comment]
RN upgrade is done so this should be solved I think

**Comment by kadiealexander** [comment]
@mallenexpensify I'm struggling a little with working out who needs to be paid and how much here, and I see you were intending on taking this over in this [comment](https://github.com/Expensify/App/issues/6876#issuecomment-1082297279).  Could you please help me make sense of the payments needed here?

**Comment by kadiealexander** [comment]
@parasharrajat could you please help me by summarising who needs to be paid with this issue? 

It seems like it's just you that is due a reporting bonus? Am I missing anything?

**Comment by kadiealexander** [comment]
@parasharrajat could you please link all the related PRs and issues here? This is a big payment and it's not a simple 1 PR = 1 payment issue like usual, so I really want to ensure I understand what happened.

**Comment by hannojg** [comment]
There is also this issue which should be fixed now:

 - https://github.com/Expensify/App/issues/12854

**Comment by kadiealexander** [comment]
@Julesssss can you please sense check the below?

Payouts due:

Issue Reporter: $250 @parasharrajat 
Contributor: n/a, handled by expert contributors
Contributor+: $16000 @parasharrajat

Eligible for 50% #urgency bonus? No

All payments to be paid via Newdot.

**Comment by kadiealexander** [comment]
Thank you @kidroca!! That's very generous of you to be willing to share your expertise with us like that. 

In that case @parasharrajat if you could please request your portion via newdot we should be good here. Thanks to everyone who worked on this!

**Comment by JmillsExpensify** [comment]
Reviewed the details for @parasharrajat. Approved for payment in NewDot based on the BZ summary above.
